### PR TITLE
Add support for visualizing HDF5 datasets.

### PR DESCRIPTION
**Pull request by jarrah42**
This contribution is split into two parts: the visualization plugin and backend changes. The client code integrates the h5web visualizer into Galaxy (https://h5web.panosc.eu). Attached is a screenshot of the visualizer.

Screen Shot 2023-01-17 at 11 02 39 AM

The backend changes add the ability to access the contents of a dataset without needing to read the whole dataset, which is essential for large files. It is based on the h5grove API (https://github.com/silx-kit/h5grove). The change adds a new dataset API for binary datatypes that implement the `get_structured_content` interface (currently only HDF5). The API is `/api/datasets/{dataset_id}/content/{content_type}` where `content_type` is one of `attr`, `meta`, `stats`, or `data`. Datasets implementing this interface are assumed to be hierarchical (see https://docs.hdfgroup.org/hdf5/develop/ for details of the HDF5 format). Parameters are used to specify where in the dataset to extract the requested information. 

## How to test the changes?
(Select all options that apply)
- [ X ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ X ] Instructions for manual testing are as follows:
  1. To test the visualizer UI you will need to upload an HDF5 dataset. It's easiest to use the `chopper.h5` file in `test-data`.
  2. Once uploaded, start the visualization tool via the `Visualize` menu or using the `Visualize` button on the dataset.
  3. After starting the tool, the various features of the tool can be tested.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by dannon** [comment]
Well that's a welcome surprise @jarrah42!  This will close #14994; great to see this.

**Comment by bgruening** [comment]
This is super cool! `make format` might help with the linting issues.

File: config/plugins/visualizations/h5web/package.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        "axios-hooks": "4.0.0",
+        "backbone": "^1.3.3",
+        "normalize.css": "8.0.1",
+        "parcel-bundler": "^1.4.1",
+        "react": "17.0.2",
+        "react-dom": "17.0.2",
+        "react-dropzone": "14.2.3",
+        "react-icons": "4.7.1",
+        "react-router-dom": "6.3.0",
+        "three": "0.141.0"
```
Comment (axelboc) [inline config/plugins/visualizations/h5web/package.json]: FYI, I think a bunch of these dependencies can be removed:

```suggestion
        "backbone": "^1.3.3",
        "parcel-bundler": "^1.4.1",
        "react": "17.0.2",
        "react-dom": "17.0.2"
```

File: config/plugins/visualizations/h5web/package.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    },
+    "scripts": {
+        "build": "parcel build src/script.js -d static"
+    },
+    "devDependencies": {
+        "@types/react": "^17.0.52",
+        "@types/react-dom": "^17.0.18",
+        "@types/react-router-dom": "5.3.3",
```
Comment (axelboc) [inline config/plugins/visualizations/h5web/package.json]:
```suggestion
```

**Comment by jarrah42** [comment]
Let me know if there's anything else I need to do. As far as I can tell the package tests are failing because the pip cache does not contain the h5grove package. I'm not sure how that gets updated. The other test failures seem unrelated to my changes.

**Comment by nsoranzo** [comment]
> Let me know if there's anything else I need to do. As far as I can tell the package tests are failing because the pip cache does not contain the h5grove package. I'm not sure how that gets updated. The other test failures seem unrelated to my changes.

I guess it depends if we want to make h5grove and orjson full dependencies (as it's been done currently) or only optional (and for dev) as it was done for `ase` in https://github.com/galaxyproject/galaxy/pull/12375 .
In the first case, you need to add `h5grove` to `packages/data/setup.cfg`, in the second to `packages/data/test-requirements.txt`

File: test/integration/test_structured_dataset.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import os
+
+from galaxy_test.base.populators import DatasetPopulator
+from galaxy_test.driver import integration_util
+
+SCRIPT_DIR = os.path.normpath(os.path.dirname(__file__))
+TEST_DATA_DIRECTORY = os.path.join(SCRIPT_DIR, os.pardir, os.pardir, "test-data")
+
+
+class StructuredDatasetTestCase(integration_util.IntegrationTestCase):
```
Comment (nsoranzo) [inline test/integration/test_structured_dataset.py]:
```suggestion
class TestStructuredDataset(integration_util.IntegrationTestCase):
```

File: lib/galaxy/dependencies/pinned-lint-requirements.txt [hunk 4]
```diff
@@ -5,6 +5,6 @@ importlib-metadata==4.2.0
 mccabe==0.7.0
 pycodestyle==2.9.1
 pyflakes==2.5.0
-ruff==0.0.228
+ruff==0.0.230
```
Comment (nsoranzo) [inline lib/galaxy/dependencies/pinned-lint-requirements.txt]: This should save you from the new Python linting errors.

```suggestion
ruff==0.0.228
```

File: config/plugins/visualizations/h5web/config/h5web.xml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        </data_source>
+    </data_sources>
+    <params>
+        <param type="dataset" var_name_in_template="hda" required="true">dataset_id</param>
+    </params>
+    <entry_point entry_point_type="chart" src="script.js" css="script.css"/>
+    <settings>
+        <input>
+            <name>explorer</name>
+            <value>false</value>
```
Comment (guerler) [inline config/plugins/visualizations/h5web/config/h5web.xml:18]: If you want to, you could also specify the type e.g. `<type>boolean</type>`.

Comment (bgruening) [reply config/plugins/visualizations/h5web/config/h5web.xml:18]: @jarrah42 this looks like a good recommendation.

**Comment by bgruening** [comment]
@jarrah42 there is a python linting error that needs to be fixed.

**Review by guerler — APPROVED** [review]
Looks very good to me. Thank you for adding this. Overall this works well for me, but we might have to revisit the `explorer` option, since the screen for some datasets does just turn blank, but that can be done in a follow-up. As a side note, you can add `<label/>` and `<help/>` tags to that input option too. Thank you so much for adding this feature. Very nice!

**Review by bgruening — APPROVED** [review]
Thats an awesome contribution @jarrah42!

Works well for my small test files, looking forward to use that on our Galaxy server on genomic HDF5 datasets :)

**Comment by bgruening** [comment]
@dannon do you think this one is ready or do we need to do something special for visualisaitons?

File: Makefile [hunk 6]
```diff
@@ -193,9 +193,6 @@ lint-api-schema: build-api-schema
 	$(IN_VENV) codespell -I .ci/ignore-spelling.txt _schema.yaml
 	$(MAKE) remove-api-schema
 
-update-navigation-schema: client-node-deps
```
Comment (dannon) [inline Makefile]: Guessing this was unintentional in the rebase?  Want me to take a pass fixing it up?

**Comment by jarrah42** [comment]
I did the original implementation using the DataProvider interface, however I found it too limited to be used to access a hierarchical dataset since it mediates access via the __iter__ interface. I guess it could be modified to make it more general purpose, but the approach I chose is more in line with REST API best practices.

File: Makefile (hunk 6 above)
Comment (jarrah42) [reply Makefile]: This seems to have come in when I merged upstream/dev into my fork, but if you can figure it out that would be appreciated!

**Pull request merged**